doc/rtd/explanation/about-cloud-config.rst

```diff
@@ -0,0 +1,167 @@
+
+The order of the ``cloud-config`` keys is unimportant and modules can be
+written in any order.
+
+.. note::
+   The :ref:`Users and Groups <mod_cc_users_groups>` module is a special case
+   where ordering matters if you want to add users to a group list.
+
+Checking your cloud-config file
+===============================
+
+Once you have constructed your cloud-config file, you can check it against
+the :ref:`cloud-config validation tool <check_user_data_cloud_config>`. This
+tool tests the YAML in your file against the cloud-init schema and will warn
+you of any errors.
+
+Example cloud-config file
+=========================
+
+The following code is an example of a complete user data cloud-config file.
+The :ref:`cloud-config example library <yaml_examples>` contains further
+examples that can be copy/pasted and adapted to your needs.
+
+.. code-block:: yaml
+
+   #cloud-config
+
+   # Basic system setup
+   hostname: example-host
+   fqdn: example-host.example.com
+
+   # Configure storage
+   storage:
+     files:
+       - path: /etc/example_file.txt
+         content: |
+           Some text to be stored in the file
+       - path: /etc/example_script.txt
+         content: |
+           #!/bin/bash
+           echo "Some text to be run in the script"
+
+   # User setup configuration
+   users:
+     - name: exampleuser
+       gecos: Example User
+       sudo: ['ALL=(ALL) NOPASSWD:ALL']
+       groups: sudo
+       home: /home/exampleuser
+       shell: /bin/bash
+       ssh_authorized_keys:
+         - ssh-rsa AAAAB3...restofpublickey user@host
+
+   # Change passwords using chpasswd
+   chpasswd:
+     exampleuser: terriblepassword12345
+
+   # Package management
+   package_update: true
+   package_upgrade: true
+   packages:
+     - git
+     - nginx
+     - python3
+
+   # Commands to run at the end of the cloud-init process
+   runcmd:
+     - echo "Hello, world!" > /etc/motd
+     - systemctl restart nginx
+     - mkdir -p /var/www/html
+     - echo "<html><body><h1>Welcome to the party, pal!</h1></body></html>" > /var/www/html/index.html
+
+   # Write files to the instance
+   write_files:
+     - path: /etc/example_config.conf
+       content: |
+         [example-config]
+         key=value
+     - path: /etc/motd
+       content: |
+         Some text that will appear in your MOTD!
+
+   # Final message, shown after cloud-init completes
+   final_message: "The system is up, after $UPTIME seconds"
+
+   # Reboot the instance after configuration
+   power_state:
+     mode: reboot
+     message: Rebooting after initial setup
+     timeout: 30
+     condition: True
+
+Ubuntu installer cloud-config
+-----------------------------
+
+For the special case where your cloud-config file will be consumed by the
+Ubuntu live installer, you will need to include the ``autoinstall:`` top level
+key. The presence of this key instructs cloud-init not to process
+the user-data itself, but instead to pass it directly to the installer for
```

Comment:
> The presence of this key instructs cloud-init not to process
the user-data itself, but instead to pass it directly to the installer for
processing.

That's not true, is it? The presence of this key doesn't prevent cloud-init from processing the user-data.

Cloud-init just ignores keys nested under the `autoinstall` key.

Reply:
good pt. cloud-init just ignores it, it doesn't take any action to "pass it to the installer"